### ISSUE TITLE
Mouse keeps hidden when the window regains focus with mouse input

### ISSUE THREAD

**Issue by SheldonZhong** [alacritty/alacritty#6703]
This is kind of related to https://github.com/alacritty/alacritty/issues/6452.

### System

OS: macOS 13.1 (22C65)
Version: 9cf2c89fd13de79b6a6999071609a55ff7e607ec

First enable the `hide_on_typing` in the config.

```
mouse:
  hide_when_typing: true
```

Steps to reproduce:
1. Launch Alacritty, let's call the window's area A.
2. Launch another window that is not Alacritty, let's call it B. Note that A and B may or may not overlap.
3. Focus on Alacritty.
4. Move the cursor to any area that is $A' \cup B$ (this is a set operation meaning "not A or B"). A Venn diagram is shown below. 
![image](https://user-images.githubusercontent.com/12372003/219108004-9b820b20-044f-4a57-8b25-77b55ff09b90.png)
5. Start typing on Alacritty. The cursor is hidden, as expected.
6. Tab to the other application using `cmd` + `tab`.
7. Move the cursor on $A \cap B'$ and click as we are trying to regain focus using the mouse.
8. Now Alacritty gets focused, but if you move around the mouse the cursor doesn't show up.

**Comment by chrisduerr** [comment]
You couldn't have phrased this issue any worse. So you're having trouble with the cursor not reappearing when the two windows overlap?

**Comment by SheldonZhong** [comment]
@chrisduerr It happens when the two windows may or may not overlap. What I was trying to say, using all the set notations, is that the bug doesn't happen when the cursor stays in the alacritty window, but not covered by the new window.

The cursor doesn't reappear if
- The cursor is outside of alacritty window. Tab to the other window, move the mouse and click the alacritty window to regain focus.
- The cursor is **inside** alacritty window, but the cursor **will be** covered by the new window, if you `cmd` + `tab` to that window. Tab to the other window, move the mouse and click to regain focus.

**Comment by SheldonZhong** [comment]
Videos included. The cursor reappears in case1.

https://user-images.githubusercontent.com/12372003/219229012-df800c4d-ed88-4694-bd1d-24e013f1e341.mp4


https://user-images.githubusercontent.com/12372003/219229023-f2c013a7-bcc3-4ca3-8758-77bd12130e4f.mp4


https://user-images.githubusercontent.com/12372003/219229032-e2103a94-a623-48ef-b12f-f496a827b972.mp4

**Comment by chrisduerr** [comment]
If you take the original patch and add `self.ctx.window().set_mouse_visible(true);` after [this](https://github.com/alacritty/alacritty/pull/6481/files#diff-c32f38542d52895124f53e3425f4243a05391d1d2bd8640f371ac6e427be5d70L1215) statement, does that fix it?

**Comment by SheldonZhong** [comment]
@chrisduerr Just tried something like
```rust
if is_focused {
    self.ctx.window().set_urgent(false);
    self.ctx.window().set_mouse_visible(true);
}
```

It doesn't work.

**Comment by chrisduerr** [comment]
If explicitly enabling the mouse again on focus doesn't work, I'm not sure what you're supposed to do.
@kchibisov any ideas? Just sounds like winit/macOS are broken.

**Comment by kchibisov** [comment]
@chrisduerr It might be winit is doing something weird here. I might try to look once I have time.

**Comment by Arzte** [comment]
I've noticed this bug when switching between virtual desktops/fullscreen on macOS, the mouse always reappears when switching desktops, however when switching back, the mouse disappears again until you click on a window.

There's a sorta work around of making sure the mouse is visible before swapping to another window, but it's obviously not a convenient one.

**Comment by bluz71** [comment]
I am also confirm that mouse visibility, with `hide_when_typing: true`, on macOS Ventura is still problematic.
I was hoping that #6481, a fix for #6452, was to be the solution.
Alas for me, no.

I am running Alacritty `0.12.0 rc2`.

I never ran #6481 just after merge; so maybe it worked, and then something else since broke it again? I don't know.

In my case I run Alacritty full screen on Desktop 1 with `tmux` and `Neovim`.
I then run a browser, Brave, in Desktop 2 to view the results of my web development (done in the Alacritty session).
I switch between Desktops 1 & 2 via three-finger swipe (constantly).
About 1 out of 5 times the mouse disappears in Alacritty, I can still select content via the trackpad, but I the mouse pointer is invisible. Hitting the `shift` restores the visibility of the pointer.

It is very sporadic.

This used to never be a problem.
I do think since I upgrade to Ventura, this has become a problem.

**Comment by bluz71** [comment]
Setting `mouse: hide_when_typing: false` avoids the issue; the cursor is always visible which is sad, but less sad than the cursor failing to re-appear after twirling the track-pad.

I am wondering if #6481 did fix #6452, but something between then and `0.12.0` re-introduced the problem?

I don't see many other macOS people raising this issue, is this limited to a few? Or are there many macOS experiencing this?

**Comment by mdwagner** [comment]
I recently upgraded to Ventura and now have this issue. I did not have this on Monterey.

**Comment by nixpulvis** [comment]
I can reproduce this on macOS 13.4.1. Looks like it might be related to this issue rust-windowing/winit#2539?

**Comment by bluz71** [comment]
I am glad I am not the only one experiencing this issue.

#6481 was supposed to have fixed this issue, but in my case not, and now for a couple others.

Hopefully the linked winit issue is the cause, and solution?

**Comment by unphased** [comment]
I can confirm this was not happening in Monterey. I updated one of my intel macbooks to Ventura and have been putting up with it. Now that I upgraded my M1 Max macbook to Ventura, the issue pops up, without me having rebuilt or otherwise modified my Alacritty install. 

The `mouse: hide_when_typing: false` config is the thing that will restore sanity to my life in the meantime. Good luck solving this bug.

**Comment by nixpulvis** [comment]
IDK if this helps, but I just had this happen to me after a Find My alert popup took focus.

**Comment by nixpulvis** [comment]
OK, now I'm seeing this happening on other macOS applications with frames the pop-over other frames within a window. So I'm willing to accept this is a macOS issue, but happy to be proven wrong.

**Comment by viccuad** [comment]
I can confirm I can reproduce this issue on Linux, alacritty 0.13.2 from NixOS 24.05.

Once the cursor disappears, it never comes back, no matter if one changes windows or moves the cursor.
Feels like a regression of https://github.com/alacritty/alacritty/issues/6452.

**Comment by kchibisov** [comment]
if you're on wayland ensure that you have cursor theme installed and it's visible to alacritty, because it sounds like you don't have it set/exposed.

**Comment by viccuad** [comment]
> if you're on wayland ensure that you have cursor theme installed and it's visible to alacritty, because it sounds like you don't have it set/exposed.

Indeed, that's it, thanks. Still unresolved on NixOS side, see https://github.com/NixOS/nixpkgs/issues/22652.

Workaround: `XCURSOR_THEME=Adwaita alacritty`.

**Comment by kchibisov** [comment]
I'm pretty sure you can set this variable globally for the user session. Anyway, with modern compositors it won't be an issue since they use wayland protocol for cursors now, which we support.

**Comment by seperman** [comment]
> if you're on wayland ensure that you have cursor theme installed and it's visible to alacritty, because it sounds like you don't have it set/exposed.

Do you mean to run `sudo apt-get install gnome-tweak-tool` @kchibisov  ?

**Comment by paddor** [comment]
I'm experiencing this on macOS `15.3.1 ` with Alacritty `Version 0.15.1 (1)`. I do have `mouse.hide_when_typing = true`, if that matters. I think I've only started experiencing this issue after upgrading to 0.15.0. After switching back to the maximized but not fullscreen Alacritty window, the mouse is hidden when hovering over the terminal content, but shown when hovering over the window title bars. It takes a few clicks or windows switching back and forth to make it work again.

**Comment by kchibisov** [comment]
Someone with macOS should go to `winit` and figure it out what's wrong. The workaround is to not hide cursor with setting, maybe there's an app that can hide it occasionally on macOS (there're plenty like that on linux).

**Comment by paddor** [comment]
@kchibisov Thanks a lot. Setting `hide_when_typing = false` looks promising.
